### ISSUE TITLE
New principle: Support common promise practices

### ISSUE THREAD

**Issue by jan-ivar** [w3ctag/design-principles#342]
**TL;DR:** Avoid limiting or specializing APIs to _"same microtask"_.

This came up in https://github.com/w3c/mediacapture-screen-share/issues/190#issuecomment-929491274 in the context of trying to specify a new platform API that should _only_ work immediately following the resolution of a promise from another platform API. A solution was proposed to limit the new API to the same microtask as the promise resolution callback, but this would have broken some fairly common practices around asynchronous programming with promises in ways that might be surprising to most people.

It would be nice to write down these common practices, and say "don't break them":

 1. Encapsulation — wrapping an async platform call in an `async function`
 1. Error logging — of a final async step (s/`return promise`/`try { return await promise } catch(e) { log.add(e) }`/)
 1. Shimming — of an async platform API in a JS library
 1. Promise branching — Multiple `then`s on the same promise `p` queue a microtask each as I understand
 2. Timeouts — `Promise.race([foo(), new Promise(r => setTimeout(r, 10000)])`

These are helpful techniques that seem benign. What they have in common is that they may trigger a microtask checkpoint, which isn't terribly obvious. I think most people would expect these techniques to work across platform APIs, so it would be nice if we could preserve that expectation. cc @annevk

**Comment by alvestrand** [comment]
Do we have a clear reference in the Web Platform for when stuff happens in a task vs a microtask?

**Comment by annevk** [comment]
I think an alternative way of phrasing OP would be to say that "A Promise represents a value, not a callback." When you attempt to overload a Promise with callback semantics you will likely run into issues.

@alvestrand I don't think we have anything yet other than that you need a task from which to do a microtask (i.e., when in parallel all you have is queue a task). At the moment we use microtasks sparingly: custom elements, mutation observers, promises (and other JavaScript jobs, details escape me at the moment), `self.queueMicrotask`, and some more questionable uses for the `img` element and "await a stable state".

In the media capture case I think there already is a clear task to use, but there is some hesitance to use it as it would involve an event or callback. That still seems like the way to go though.

**Comment by atanassov** [comment]
@jan-ivar, reading through the conversation we're having difficulty deciding if you're suggesting a principle that applies to Promises only (in which case that would go to our [promises guide](https://www.w3.org/2001/tag/doc/promises-guide)) or a more general design principle that goes beyond?

**Comment by alvestrand** [comment]
I think this pattern (doing something that was only possible before a particular task has finished) was first introduced by @youennf in https://w3c.github.io/webrtc-encoded-transform/#stream-creation - Apple's proposed new API to transform streams for encoded video frames - where you can only create a stream by having set a controller before the task enqueued by the stream creation executes.

There are multiple ways of specifying this "queue a task to shut the door" type of operation. The particular one that triggered this discussion is https://wicg.github.io/conditional-focus/#focusable-media-stream-track

**Comment by youennf** [comment]
There are examples predating WebRTC encoded transform, say service worker FetchEvent.respondWith.
Encapsulating "queue a task to shut the door" operations in an event handler is known and does not seem to cause issues, ditto with a task scope.
Narrowing down to microtasks is more rare, there should be a real justification to select this level. Maybe it should be discouraged.

Looking back to the initial issue at https://github.com/w3c/mediacapture-screen-share/issues/190, the desired scope is probably at the level of task/event handler/callback, not microtask.

**Comment by martinthomson** [comment]
@jan-ivar, the TAG is struggling a little with this one.  The framing of this question seems to be based on there being something going on here that we generally avoid.

If I understand, you would prefer that if we pass an argument to a promise-returning function, there is a reasonable expectation that the argument is valid until the promise is resolved.

However, in some of the cases highlighted, there is a desire to set a time limit on validity.  Trying to align that with task/microtask boundaries seems wrong to me also, because things like audio/video encoding are time-based, not task-based.  Perhaps there is an angle there that we can push on.

It sounds like there is some disagreement still, at least when the last messages on this issue were exchanged.  Has the discussion in the working group moved on?

Either way, do you think that you could draft a principle on this, @jan-ivar ?  It might not be an absolute thing, but we might be able to put tighter constraints on what people do.  Perhaps these encoding cases will still need to do some ugly stuff, but at least we'd be able to have things be a little better structured.

**Comment by jan-ivar** [comment]
The discussion concluded satisfactorily. I'll add a PR.
